File: maxChunksToSorted.cpp
/**
 * @author f4prime
 * @email zzz136454872@163.com
 * @aim a plain model for leetcode
 */

#include<iostream>
#include<string>
#include<vector>
#include<algorithm>
#include<map>
#include<stack>
#include<queue>

using namespace std;

//显示中间结果
void print(vector<int> a)
{
    for(int i=0;i<(int)a.size();i++)
        cout<<a[i]<<" ";
    cout<<endl;
}

void print(vector<vector<int>> a)
{
    for(int i=0;i<(int)a.size();i++)
    {
        for(int j=0;j<(int)a[i].size();j++)
            cout<<a[i][j]<<" ";
        cout<<endl;
    }
}

void print(vector<bool> a)
{
    for(int i=0;i<(int)a.size();i++)
    {
        cout<<i<<" ";
        if(a[i])
            cout<<"true ";
        else
            cout<<"false ";
    }
    cout<<endl;
}

void print(bool a[],int len)
{
    for(int i=0;i<len;i++)
    {
        cout<<i<<" ";
        if(a[i])
            cout<<"true ";
        else
            cout<<"false ";
    }
    cout<<endl;
}

//#define testMod

#ifdef testMod
void test()
{
    
}
#endif

#ifndef testMod
class Solution {
public:

    int maxChunksToSorted(vector<int>& arr) {
        int out=0;
        int m=0;
        for(int i=0;i<(int)arr.size();i++)
        {
            m=m>arr[i] ? m:arr[i];
            if(m==i)
                out++;
        }
        return out;
    }
        
private:
};



#endif

int main()
{
#ifdef testMod
    test();
#endif

#ifndef testMod
    Solution sl;
    vector<int> in={1,0,2,3,4};
    cout<<sl.maxChunksToSorted(in)<<endl;


    
#endif 
    return 0;
}

File: maxChunksToSorted.py
from typing import List


# 768. 最多能完成排序的块 II
# 769. 最多能完成排序的块
class Solution:
    def maxChunksToSorted(self, arr: List[int]) -> int:
        def same(a, b):
            try:
                for k in a:
                    if a[k] != b[k]:
                        return False

                return True
            except:
                return False

        count = 0
        arr2 = arr.copy()
        arr2.sort()

        a = {}
        b = {}

        for i in range(len(arr)):
            a[arr[i]] = a.get(arr[i], 0) + 1
            b[arr2[i]] = b.get(arr2[i], 0) + 1

            if same(a, b):
                a = {}
                b = {}
                count += 1

        return count


arr = [5, 4, 3, 2, 1]
arr = [2, 1, 3, 4, 4]
print(Solution().maxChunksToSorted(arr))
